api/src/main/java/org/apache/iceberg/catalog/Catalog.java

```diff
@@ -303,4 +303,83 @@ default boolean dropTable(TableIdentifier identifier) {
    * @throws NoSuchTableException if the table does not exist
    */
   Table loadTable(TableIdentifier identifier);
+
+  /**
+   * Instantiate a builder to either create a table or start a create/replace transaction.
+   *
+   * @param identifier a table identifier
+   * @param schema a schema
+   * @return the builder to create a table or start a create/replace transaction
+   */
+  default TableBuilder buildTable(TableIdentifier identifier, Schema schema) {
+    throw new UnsupportedOperationException("Not implemented: buildTable");
+  }
+
+  /**
+   * A builder used to create valid {@link Table tables}.
+   * <p>
+   * Call {@link #buildTable(TableIdentifier, Schema)} to create a new builder.
+   */
+  interface TableBuilder {
```

Comment:
I'll need to fix CachingCatalog.

Reply:
Also, we will need to add extra validation to the builder in `HadoopCatalog`.

Reply:
Okay, I added validation for `HadoopCatalog` but we can add support for `buildTable` in `CachingCatalog` later. Looks like we will need a separate builder implementation that would delegate to another builder but add caching callbacks.